### ISSUE TITLE
[PROPOSAL] OpenSearch JS Version 3.0 and OpenSearch Typescript

### ISSUE THREAD

**Issue by nhtruong** [opensearch-project/opensearch-js#803]
As a part of our effort to generate all clients from the OpenSearch API Spec, and modernize the Node.js client, we are planning some major updates for this repo.

## OpenSearch JS Version 3.0
The next major version of this client is going to be fully generated from the API Spec. The vast majority of the current client is hand-written with a lot of inconsistency in the implementations of the less known and less used features:
- Support of both camelCased and snake_cased query string parameters
- Support of both camelCased and snake_cased api method names
- Overriding the HTTP Method of an API Function

Said features are going away in 3.0 to promote a cleaner and more efficient code base:
- The accepted query string parameters will now be the same as what the OpenSearch cluster expects, which is in snake_case. The client will no longer loop through the list of parameters passed to an API method to convert the camelCased keys into snake_case keys every time an API function supporting this feature is invoked.
- The client will only support camelCased API module and function names as this is a JavaScript convention.
- The client will not allow the HTTP method of an API function to be overridden by passing the `method` param to the function. That means, usages of this feature will result in error from the server as the `method` param will be treated as a query string parameter.

Version 3.0 will also introduce other improvements:
- Lazy loading of API modules and functions: Instead of loading every API function as soon as this node package is required, those functions will be loaded on the fly as they are invoked for the first time.
- More readable and maintainable file and folder structure:
  - Each API function will now be in their own file: Currently, only functions of the core API namespace are in their own files, while functions of every other namespaces are all written in the same giant `.js` file.
  - API functions will be divided into folders: Instead of having all API function and module files in one flat folder, the API function files mentioned above will be put into their respective namespace folders.

## OpenSearch Typescript
Once OpenSearch JS 3.0 is released, we will start rewriting this client in Typescript in a different repo. This rewrite will carry over all improvements introduced in OpenSearch JS 3.0, along with some other changes to modernize and streamline the client:
- Dropping support for CommonJS in favor of ES6
- Instead of callback, the API functions will be written using `async/await` syntax.
- [Tentative] You will also no longer be able to pass the transport options to every API function call. This is a niche feature that can be replaced by instantiating multiple client instances with different transport options. A better option is adding the ability to modify a client instance's transport options before an API call.
- A complete rewrite of the transport layer to get rid of many duct-tape solutions in the current JS client.
- More streamlined integration of AWS Sigv4

The Typescript client will be a completely different client, and we're aiming to do it properly with ample research. OpenSearch JS 3.x will be constantly be updated when OpenSearch Typescript is under construction, and it will continue receive API updates long after OpenSearch Typescript is released to give the user time to switch over.

If you have any concern or suggestion, feel free to comment below.

**Comment by dblock** [comment]
I love the end goal: an opensearch-js to be rewritten in TypeScript, and to use a code generator.

> Once OpenSearch JS 3.0 is released, we will start rewriting this client in Typescript in a different repo. 

If I understand correctly the TypeScript client will be completely compatible with opensearch-js 3.0, so it should be a branch and eventually be merged to main as a replacement, then we would be shipping a non-breaking opensearch-js 3.x, but rewritten in TypeScript.

If this is not possible, I don't see why we need openserach-js 3.0 at all. It's a breaking change for users, then they would have another breaking change switching to a TypeScript client, so why cause the pain twice?

> * The accepted query string parameters will now be the same as what the OpenSearch cluster expects, which is in snake_case.

I wonder whether we can implement an optional compatibility mode for this that users would inject in the client to preserve backwards compat.

**Comment by nhtruong** [comment]
> If this is not possible, I don't see why we need opensearch-js 3.0 at all. It's a breaking change for users, then they would have another breaking change switching to a Typescript client, so why cause the pain twice?

The breaking changes in 3.0 will only affect a few applications that use those very obscured features that are implemented sporadically (not all API functions support them):
- Per Javascript convention, module and function names are camelCased. So, most JS apps will not use the snake_cased version.
- Our JS docs only mention snake_cased query string params. The fact that some functions accept the camelCased params is a hidden feature.
- Overriding HTTP method is not documented either, and it's not at all useful (if the user needs to make custom requests not provided by our API functions, they should use `client.http` instead).

So, the switch to 3.0 will be straight forward for most users.

The Typescript client, however, will affect all users as we're also rewriting the transport layer. The options you provide during client instantiation will be different. Legacy apps will also be affected by the use of modern `async/await` syntax (and get rid of callbacks). The API function calls will remain the same (creating an index named book will still be `client.indices.create({ index: 'books' }`) unless the user made use of callbacks in their app. The switch from 3.x to TS will also only happen in very few places for modern apps.

Having both 3.x and TS clients existing in parallel affords the user some time to modernize their apps.

> I wonder whether we can implement an optional compatibility mode for this that users would inject in the client to preserve backwards compat.

It's possible, but it's not straight forward. This feature is also a performance issue since it [loops through the list of params](https://github.com/opensearch-project/opensearch-js/blob/main/api/utils.js#L51-L54) to reconstruct the querystring params object for every API call. We will also run into a complication with the `notifications` namespace where `query.string.names.like.this` is introduced. It's a lot of work to preserve a lesser used feature that's actually causing more harm than good.

**Comment by dblock** [comment]
> Having both 3.x and TS clients existing in parallel affords the user some time to modernize their apps.

This can be achieved by shipping older versions from a 2.x branch. I think it's a small but important nuance whether we use two separate repos or not.

**Comment by nhtruong** [comment]
> I think it's a small but important nuance whether we use two separate repos or not.
Our release pipeline always use what's in the main branch. So, having 2 different release pipelines in the same repo is not possible unfortunately. Moreover, since the TS client is a complete rewrite, it will have its own list of issues, and having these issues in a separate repo is easier to manage.

**Comment by nhtruong** [comment]
I've just uploaded the JS Client 3.0 Preview into this [feature branch](https://github.com/opensearch-project/opensearch-js/tree/api_3.0). It was completely generated from the OpenSearch API Spec, and it is fully functional (The types are still from 2.x. Working on generating the types from the spec next). It has all the changes and improvements mentioned in the OP. I also refactored a lot of the code:
- [Old index.js](https://github.com/opensearch-project/opensearch-js/blob/main/api/index.js) vs [New index.js](https://github.com/opensearch-project/opensearch-js/blob/api_3.0/api/index.js)
- [Old CatApi](https://github.com/opensearch-project/opensearch-js/blob/main/api/api/cat.js) vs [New CatApi](https://github.com/opensearch-project/opensearch-js/blob/api_3.0/api/cat/_api.js)
- [Old snapshot.clone](https://github.com/opensearch-project/opensearch-js/blob/main/api/api/snapshot.js#L151-L215) vs [New snapshot.clone](https://github.com/opensearch-project/opensearch-js/blob/api_3.0/api/snapshot/clone.js#L39-L55)

**Comment by nhtruong** [comment]
Another breaking change we should consider for 3.0 is ending support for older Node.js versions. The support of old Node.js (We current support Node 10 and up) is preventing us from upgrading dependency packages with [tsd](https://github.com/opensearch-project/opensearch-js/pull/806) is the newest example, and it also poses a security risk. (Node 16 stopped receiving security updates on [Sep 11, 2023](https://endoflife.date/nodejs))

**Comment by nhtruong** [comment]
Just had a discussion with @timursaurus on the TS client:
- The transport layer, which only receives a handful of update a year, can be its own Node package to reduce wasted bandwidth for the user when most of the client updates happen in the API. Though do note that the transport layer is very small comparing to the API, and the size gap is only growing as more API functions are being added. So, the bandwidth saving might not be worth it for the extra complexity. We need do to more research into this topic. 
- Potentially make the TS client backward compatible with CommonJS with a transpiler (This is of a lower priority)
- See if we can avoid using 3rd-party packages, like axios, to make HTTP requests, and use native Node packages instead. Overall reduce the number of dependencies for the new clients.
- Only support AWS SDK 3 for Sigv4.

@timursaurus is going to research the current transport layer to see what business logic should be carried over to the new client. He'll work on the POC and publish his findings for discussion in a different issue in this repo. I'll modify the API generator (which is still WIP) for JS 3.0 to build the API layer for the TS client.

**Comment by nhtruong** [comment]
Updates on type generation:
We've gotten through the hardest part of type generation. Check out the generated type files in the [./types](https://github.com/opensearch-project/opensearch-js/tree/api_3.0/types) folder. I've divided types into 2 folders:
- [namespace type files](https://github.com/opensearch-project/opensearch-js/tree/api_3.0/types/namespaces): Each file has the input type (Request) and output type (Response)
- [component type files](https://github.com/opensearch-project/opensearch-js/tree/api_3.0/types/components): Each file contains a collection of component types of a certain category.

Similar to the main code, in 3.0, types are also divided into more readable files instead of a handful of giant files.

Remaining work regarding type generation:
- Handle edge cases like [RequestProcessor](https://github.com/opensearch-project/opensearch-js/blob/api_3.0/types/components/search_pipeline._common.d.ts#L52-L63)
- Handle bugs like when `tasks.list`'s `group_by` query type is [explicitly defined](https://github.com/opensearch-project/opensearch-js/blob/api_3.0/types/namespaces/tasks.list.d.ts#L23) instead of referencing [GroupBy](https://github.com/opensearch-project/opensearch-js/blob/api_3.0/types/components/tasks._common.d.ts#L18) type.
- Generate Client type (This is the type that actually ties everything together)

**Comment by nhtruong** [comment]
Happy to announce that we're ready for `3.0.0.beta` release on Monday (US Time). Check out the [api_3.0 branch](https://github.com/opensearch-project/opensearch-js/tree/api_3.0) if you want to test it out yourself. Note that instead of 
```js
const { Client } = require('@opensearch-project/opensearch');
```
you will need to use 
```js
const { Client } = require('./index'); // be mindful of where your test file is relatively to the `index.js` file in the root of the branch
```
to load the source package. The `Client` class then can be used nearly as identical as you did in 2.X minus some breaking changes already discussed above.

On top of the issues mentioned in the [previous update](https://github.com/opensearch-project/opensearch-js/issues/803#issuecomment-2246381426), the typing for the API have been completely overhauled. There's now only 1 type definition for [Client](https://github.com/opensearch-project/opensearch-js/blob/api_3.0/types/Client.d.ts). This will be a huge quality-of-life (QoL) change from 2.X where we had 3 different type definitions for Client. The generated Client type also now correctly reflects whether `params` is required for each API function, among other QoL updates I added while working on the overhaul.

Would love to have some input from the dashboard team as well especially after the beta version is released.
@AMoo-Miki

**Comment by dblock** [comment]
@nhtruong Excited to try it! Can we do 3.0.0.beta1 or will the next beta update that tag? I am sure it won't be the first and last beta ;)

**Comment by nhtruong** [comment]
Just tested this myself on my IDE ([JetBrains WebStorm](https://www.jetbrains.com/webstorm/)). Typing works on both 2.X and 3x, and on both `ts` and `js` files. The main difference is the 3.X and Typescript combo provides the best code hints and autocomplete:
- 3.x and CommonJS: The IDE handles the requests accurately, but autocomplete for the response body is the combination of all properties of all response bodies defined in 3.X even though it accurately determines the type of the response body (Seems like a bug in the IDE).
- 3.x and Typescript: The IDE handles both the requests and responses accurately.
- 2.x (both js and ts): The IDE handles both the requests accurately (The types are of much lower resolution than 3.x), but all response bodies are the generic `Record<string, any>`, which is sometimes incorrect.

I've taken some screenshots to compare between 3.x and 2.x (both in Typescript) for this code snippet:
```ts
const Client3 = require('opensearch_3').Client; // 3.0.0-beta.1
const Client2 = require('opensearch_2').Client; // 2.11.0

const clientOptions = { node: 'http://localhost:9200' } // simplified
const client3 = new Client3(clientOptions);
const client2 = new Client2(clientOptions);

const start = async function() {
  const nodes_info_3 = (await client3.nodes.info({ metric: [] })).body;
  const search_3 = (await client3.search({})).body;
  const cat_indices_3 = (await client3.cat.indices({ format: 'json' })).body;
  console.log(nodes_info_3);
  console.log(search_3);
  console.log(cat_indices_3);

  const nodes_info_2 = (await client2.nodes.info({ metric: [], })).body;
  const search_2 = (await client2.search({})).body;
  const cat_indices_2 = (await client2.cat.indices({ format: 'json' })).body;
  console.log(nodes_info_2);
  console.log(search_2);
  console.log(cat_indices_2);
}
start();
```

Both 3.x and 2.x have good coverage for API function params:

![nodes_info_3](https://github.com/user-attachments/assets/b366919d-a9bb-4bfb-b3de-a547e094579a)
![nodes_info_2](https://github.com/user-attachments/assets/d1e64477-32db-40e0-ad7b-2a123841e2e1)

---

However, types in 3.x has much higher resolutions. When you use `client.nodes.info()`, the IDE helps you fill out the metrics supported by the endpoint in 3.x but not in 2.x:

![metric_3](https://github.com/user-attachments/assets/eba3b7d7-642c-4404-821f-5b4cc8271464)
![image](https://github.com/user-attachments/assets/6af1e086-63e1-4de2-b5d3-bc0ed4aa0ff6)

---

The response bodies in 3.x are also of much higher resolution and more accurate. Response body for `search()`:

![response search_3](https://github.com/user-attachments/assets/7cfea9e7-df11-4e91-b272-a4295ed1c4c6)
Notice how in 3.x above you get all the properties of a search response body, while in 2.x below, it's just a generic Object.
![response search_2](https://github.com/user-attachments/assets/7860493f-b6af-4bc1-90e9-0cacc8399816)

---

The each cat endpoint (when `format: 'json'` is used) returns an array of objects, however 2.x assumes it's `Record<string, any>` like all other response bodies,

![response cat_2](https://github.com/user-attachments/assets/46425668-624a-439e-9074-7ff45f20469c)

 while 3.x correct determines that it's it's an array:
 
![response cat_3a](https://github.com/user-attachments/assets/b1ae106e-7820-4762-9d5d-19774f4ab5ec)

and the data type of the items are of a very high resolution: 

![response cat_3b](https://github.com/user-attachments/assets/d18b2199-6df1-4801-bbd2-23255ae5180f)

**Comment by dblock** [comment]
I am with you - VScode is having a separate issue that hasn't changed with 3.x, but we should fix it anyway. https://github.com/opensearch-project/opensearch-js/issues/839

**Comment by nhtruong** [comment]
@dblock The use of generic prevents VScode from properly determine the actual body of a response. While JetBrains can property determine the type, its `inspect source` feature also points you to the generic definition instead of the actual response body. I'm going to get rid of generic in `beta.2`. While I'm at it, I'm also going to restructure the request/response type definitions so that they are importable, and have more meaningful names instead of `Response` or `ResponseBase` like what we see below

![image](https://github.com/user-attachments/assets/5e23fe8c-0e55-4870-a5db-77a1b498f853)

**Comment by nhtruong** [comment]
Releasing [beta.2](https://github.com/opensearch-project/opensearch-js/issues/840) where the response types are no longer generic parameters for `ApiResponse`. Instead they are now their own types that extend `ApiResponse`. This allows Webstorm to correctly `Go to -> Declaration`. 

However, in CommonJS (i.e. using `require` instead of `import`), the IDE still can't properly provide correct autocomplete for response body objects: It still shows all possible response properties. All response type declarations, although placed in different modules, still share the same name.  I speculated that the IDE assumes a response to be any of the declaration for autocomplete and combines of their properties together in CommonJS (EVEN THOUGH its `Go to -> Declaration` can pinpoint the correct delcaration). This does not happen in Typescript files thankfully. Manually renaming a tested response declaration to something unique resolved this for Webstorm.

Goals for `beta.3`:
- Restructure Request/Response type declaration files to allow the user to import them (this is a big lift)
- Give Request/Response type declarations globally unique names

**Comment by dblock** [comment]
Beta 2 looks better with `import`!

<img width="560" alt="Screenshot 2024-08-14 at 3 30 59 PM" src="https://github.com/user-attachments/assets/23424470-84d2-4471-8b92-19bf2fef6ab3">

<img width="610" alt="Screenshot 2024-08-14 at 3 32 07 PM" src="https://github.com/user-attachments/assets/3db59ace-f90c-472e-9864-fce0e4f51fbe">

**Comment by dblock** [comment]
I had search written like so, this doesn't work anymore because of `director: 'miller'` (`director` is a field in the index mapping).

<img width="758" alt="Screenshot 2024-08-14 at 3 36 24 PM" src="https://github.com/user-attachments/assets/3b92a1db-263b-4ea2-93c5-9da576da9842">

**Comment by nhtruong** [comment]
@dblock found the issue in the generator. The fix will be in `beta.3`

**Comment by nhtruong** [comment]
releasing [beta.3](https://github.com/opensearch-project/opensearch-js/issues/841) where all Request and Response objects have unique names, and the fix for [this](https://github.com/opensearch-project/opensearch-js/issues/803#issuecomment-2289728357). It resolves the ambiguity for WebStorm when determining which Request and Response objects to used for autocomplete in CommonJS files. @dblock, let me know if you find any issues with VSCode in beta.3

---
There's one last QoL change that I'm working on in `beta.4`. In `beta.3`, if you want to import the request types for your IDE to enforce type checking, you have to write out the api name twice (in the imported object name and the file to import) and have to import these objects one by one:
```ts
import type { Search_RequestBody } from 'opensearch_3/types/functions/search';
import type { Nodes_Info_Request } from 'opensearch_3/types/functions/nodes.info';
import type { CreatePit_Request} from 'opensearch_3/types/functions/create_pit';

...

const createPit_params: CreatePit_Request = { allow_partial_pit_creation: true, index: [movies] }
const createPit = await client.createPit(createPit_params)
console.log(createPit.body.pit_id);

const nodes_info_params: Nodes_Info_Request = { metric: ['jvm'] }
let nodes_info = await client.nodes.info(nodes_info_params);
console.log(nodes_info.body.nodes[0].jvm.mem.direct_max_in_bytes);

const search_body: Search_RequestBody = { query: { match: { director: 'miller' } } };
const search = await client.search({ index: movies, body: search_body });
console.log(search.body.hits.hits[0]._source);
``` 

In `beta.4` I'm aiming for the below, which is similar to what `2.x` is doing:
```ts
import type * as API from 'opensearch_3/types/api';

...

const createPit_params: API.CreatePit_Request = { allow_partial_pit_creation: true, index: [movies] }
const createPit = await client.createPit(createPit_params)
console.log(createPit.body.pit_id);

const nodes_info_params: API.Nodes_Info_Request = { metric: ['jvm'] }
let nodes_info = await client.nodes.info(nodes_info_params);
console.log(nodes_info.body.nodes[0].jvm.mem.direct_max_in_bytes);

const search_body: API.Search_RequestBody = { query: { match: { director: 'miller' } } };
const search = await client.search({ index: movies, body: search_body });
console.log(search.body.hits.hits[0]._source);
```

---

I also found another [issue in the spec](https://github.com/opensearch-project/opensearch-api-specification/issues/514) while testing `beta.3`. The `create_pit` endpoint, for example, only accepts an array of index per the spec, even though it should also accept a string as well. This will become a breaking change for users passing in a string for the index param:
```ts
await client.createPit({ index: 'movies' }) // This will cause a typescript error
await client.createPit({ index: ['movies'] }) // This is the only accepted format
```
@Xtansia is this also a problem for the JAVA client?
@dblock should we hold off the 3.0.0 release until this issue is resolved in the spec?

**Comment by dblock** [comment]
> releasing [beta.3](https://github.com/opensearch-project/opensearch-js/issues/841) where all Request and Response objects have unique names, and the fix for [this](https://github.com/opensearch-project/opensearch-js/issues/803#issuecomment-2289728357). It resolves the ambiguity for WebStorm when determining which Request and Response objects to used for autocomplete in CommonJS files. @dblock, let me know if you find any issues with VSCode in beta.3

Works well with `import`, `search` and other APIs look nice. Doesn't with `require`, but I believe that wasn't expected?

![Screenshot 2024-08-15 at 2 43 05 PM](https://github.com/user-attachments/assets/a5a9a147-038a-4037-b880-b780f993d211)

**Comment by nhtruong** [comment]
Releasing [beta.4](https://github.com/opensearch-project/opensearch-js/issues/842):
- Importing Request and Response types has been streamlined:
```ts
import { Client, type API } from '@opensearch-project/opensearch'

const createPit_params: API.CreatePit_Request = { allow_partial_pit_creation: true, index: [movies] }
const nodes_info_params: API.Nodes_Info_Request = { metric: ['jvm'] }
const search_body: API.Search_RequestBody = { query: { match: { director: 'miller' } } }
```
- Each api function `.js` file is now accompanied by a `.d.ts` file in the same folder that contains its Request and Response types. For example, the [knn](https://github.com/opensearch-project/opensearch-js/tree/api_3.0/api/knn)  namespace folder now has the following files:
  - _api.js
  - delete_model.d.ts
  - delete_model.js
  - get_model.d.ts
  - get_model.js
  - search_models.d.ts
  - search_models.js
  - stats.d.ts
  - stats.js
  - train_model.d.ts
  - train_model.js
  - warmup.d.ts
  - warmup.js

- Another improvement over 2.x: `OpensearchAPI` and `Client` are now in their own `.js` files with matching `.d.ts` files. `OpensearchAPI` is now defined as a class instead of a constructor function, matching its declared type. Overall, the file/folder structure of 3.0.0 is a lot more coherent than 2.x

**Comment by andreafspeziale** [comment]
Hello and sorry to jump in with a kind of dumb question/point... I'm wrapping this client in a package for NestJS applications and I'm exporting all your types for convenience. I couldn't find any way to type search search responses:

```ts
const s = (await this.osClient.search(q)).body.hits.hits.map(
      (el) => el._source,
    );
```

`el._source` would be of type `Hit._source?: Record<string, any> | undefined`

Are you planning to use generics in the new codebase in order to enable such use-case? Or am I already missing something in the current implementation?

**Comment by nhtruong** [comment]
@andreafspeziale 
Do you mean `el._source` is supposed to be the generic representing the doc structure of an index instead of `Record<string, any>` like in [this](https://github.com/opensearch-project/opensearch-js/blob/main/api/types.d.ts#L1001)? 

Unfortunately what see now was generated directly from the [OpenSearch API Spec](https://github.com/opensearch-project/opensearch-api-specification). As of right now, the spec doesn't have any directive to specify that a certain property is a generic parameter. So instead of
```ts
const s = (await this.osClient.search<MyDocument>(q)).body.hits.hits.map(
  (el) => el._source,
);
```
Now you'll do
```ts
const s = (await this.osClient.search<MyDocument>(q)).body.hits.hits.map(
  (el) => el._source as MyDocument,
);
```

![image](https://github.com/user-attachments/assets/6010e293-5918-475b-a5a5-ce7716f1bca2)

For clarification, the generics we got rid of was the body property of the [ApiResponse](https://github.com/opensearch-project/opensearch-js/blob/main/lib/Transport.d.ts#L89) interface. Instead, we define every response as an extension of ApiResponse the replace the body with the response body of the endpoint like in [search](https://github.com/opensearch-project/opensearch-js/blob/api_3.0/api/_core/search.d.ts#L112) for example.

**Comment by dblock** [comment]
> As of right now, the spec doesn't have any directive to specify that a certain property is a generic parameter.

There aren't many of those, maybe we can keep a workaround in the generator?

**Comment by nhtruong** [comment]
> There aren't many of those, maybe we can keep a workaround in the generator?

We could but It won't be a quick fix unfortunately. The generic is actually nested in several layers:
`client.search<T> -> Search_Response<T> -> body<T> -> hits<T> -> hits<T> -> _source: T`. In other words, `_source` being generic bubbles up all the way to the API function signature. We'll need to rework the `JSON Schemas -> Typescript Types` component, and add the Generic feature to the client signature renderer too.

**Comment by andreafspeziale** [comment]
> This brings up another question: Should we hold up 3.0.0 to work on this feature (It could take awhile), or save this for 3.1.0 because the user actually have a workaround to tell TS that _source is of a certain type.

@nhtruong I would save it for later.

P.S: I moved to `3.0.0-beta.4`, is the `Client.d.ts` broken? Can't see the client methods:

<img width="980" alt="Screenshot 2024-08-20 alle 00 37 19" src="https://github.com/user-attachments/assets/7f391ae8-62f8-462c-a5b4-a4a5e0445c66">

Those are all the options I can see, it seems I can't resolve `OpenSearchAPI` which comes from `declare class Client extends OpenSearchAPI` (`Client.d.ts`)

**Comment by nhtruong** [comment]
`Client extends OpensearchAPI` which has all the API methods. However, the `OpenSearchAPI` files (both `.js` and `.d.ts`) are not included in beta.4 for some reason:
![image](https://github.com/user-attachments/assets/c5ab9b0c-fbe7-4130-abbe-c1ea30e25ebe)
Even though they exist in the source:
![image](https://github.com/user-attachments/assets/fa7b5a8c-9421-4a3b-af2c-2f228616112d)
Not sure how that happened.

**Comment by nhtruong** [comment]
Release [beta.5](https://github.com/opensearch-project/opensearch-js/issues/846):
Replacing `.npmignore` with `package.json#files` that includes: `"api/", "lib/", "index.d.ts", "index.js", "index.mjs", "README.md", "LICENSE.txt"` to:
- Fix the missing files issue in `beta.4`
- Reduce memory/bandwidth footprint for the package (since we've already put a lot of work into making all modules and api function lazy load 😀)

**Comment by nhtruong** [comment]
@andreafspeziale how did you use generic for `search` in 2.x? Currently we have 2 type declaration for Client in 2.x:
- in [./index.d.ts](https://github.com/opensearch-project/opensearch-js/blob/main/index.d.ts#L4812-L4841), which is the default but doesn't utilize TDocument generic (It has other generics for request and response)
- in  [./api/new.d.ts](https://github.com/opensearch-project/opensearch-js/blob/main/api/new.d.ts#L1918-L1933) where TDocument generic is actually defined used.

So to actually use the client type declaration with TDocument generic in 2.11, it's rather convoluted:
```ts
import { Client } from '@opensearch-project/opensearch';
import { Client as NewClient } from '@opensearch-project/opensearch/api/new';

const client = (new Client2({...})) as unknown as NewClient; // have to coerce it to unknown first
const search = (await client.search<MyDocument>({...})).body;
const doc = search.hits.hits[0]._source; // Then TS will see doc as MyDocument
```

**Comment by andreafspeziale** [comment]
@nhtruong Sorry for the late response 😞 I'm not using 2.x in my projects, I'm already on `^3.0.0-beta.5`.
You can check them out here:

- https://github.com/andreafspeziale/nestjs-search
- https://github.com/andreafspeziale/os-cli

(The second is leveraging the first)

Drop a ⭐ if you found them interesting! Have been developed with ❤️ in my free time

**Comment by nhtruong** [comment]
@andreafspeziale Wow those are great projects. And thanks for helping us test the beta.
We will release beta.7 soon with even more endpoints added from the spec.

**Comment by andreafspeziale** [comment]
No problem @nhtruong! Thx for your hard work people!

**Comment by nhtruong** [comment]
Closing as this as 3.0.0 has been release release. Thanks everyone for chiming. Please don't hesitate to open an issue if you find any problem with 3.0.0.
